### PR TITLE
feat(Core/Hooks): Add some script hooks for custom module.

### PR DESCRIPTION
**Pull request by eyeofstorm**
## Changes Proposed:

- Implement a hook called after player move worldport. (teleport between two maps)
- Implement a hook called after player move teleport. (teleport within map)
- Implement a hook called from beginning of creature Update tick.
- Implement a hook called after creature finished cast a spell.

## Tests Performed:

- Compiled with test module 'mod-npc-bots' at Windows 10, ubuntu and macOS(Big Sur).
- Summon a bot in game, then attack a creature. make sure hook 'OnSpellGo' called correctly.
- Player teleport to another map, make sure hook 'OnPlayerMoveWorldport' called correctly.
- Player teleport to another area in the same map, make sure hook 'OnPlayerMoveTeleport' called correctly.

## How to Test the Changes:

1. Try hooking OnPlayerMoveWorldport(Player* player) , make sure everything is working.
2. Try hooking OnPlayerMoveTeleport(Player* player) , make sure everything is working.
3. Try hooking OnBeforeCreatureUpdate(Creature*, uint32) , make sure everything is working.
4. Try hooking OnSpellGo(Unit const* caster, Spell const* spell, bool ok) , make sure everything is working.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Entities/Creature/Creature.cpp [hunk 1]
```diff
@@ -587,6 +587,13 @@ bool Creature::UpdateEntry(uint32 Entry, const CreatureData* data, bool changele
 
 void Creature::Update(uint32 diff)
 {
+    //mod-npc-bots
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Creature/Creature.cpp:590]: delete this comments

Comment (eyeofstorm) [reply src/server/game/Entities/Creature/Creature.cpp]: > //mod-npc-bots

all comments above has been deleted.

File: src/server/game/Spells/Spell.cpp [hunk 2]
```diff
@@ -3622,6 +3622,8 @@ void Spell::_cast(bool skipCheck)
             SendCastResult(castResult);
             SendInterrupted(0);
 
+            sScriptMgr->OnSpellGo(m_caster->ToCreature(), this, false);
```
Comment (Winfidonarleyan) [inline src/server/game/Spells/Spell.cpp:3625]: for this hook no need 1-st arg, you can extract this with `GetCaster()->ToCreature()`

Comment (eyeofstorm) [reply src/server/game/Spells/Spell.cpp]: remove unessecary 1st parameter of hook OnSpellGo

**Comment by LukasVolgger** [comment]
Hello, where can we find the module?

**Comment by eyeofstorm** [comment]
> Hello, where can we find the module?

you can find my test module here.
[https://github.com/eyeofstorm/mod-npc-bots](https://github.com/eyeofstorm/mod-npc-bots).

**Comment by LukasVolgger** [comment]
Nice, thx! Would you like to list it in the modules catalouge of AC?

**Comment by eyeofstorm** [comment]
> Nice, thx! Would you like to list it in the modules catalouge of AC?

This module is still in development. so it's not yet time to list it in the cataloug.

**Comment by LukasVolgger** [comment]
Np, let us inow if you need tests or smthg.

**Comment by Nyeriah** [comment]
@eyeofstorm you no longer need this?

**Comment by eyeofstorm** [comment]
> 
yes, I want to rewrite my module so no need at this moment. maybe reopen it later.

**Pull request closed without merging**